Add new order ratio being added to 8.0.0 (#16871)
pingcap/docs: system-variables.md

- For long-term stable OLTP workload, or if the user is affirmative on the existing execution plans, it is recommended to use the `determinate` mode to reduce the possibility of unexpected execution plan changes. Additionally, you can use the [`LOCK STATS`](/sql-statements/sql-statement-lock-stats.md) to prevent the statistics from being modified and further stabilize the execution plan.

### tidb_opt_ordering_index_selectivity_ratio <span class="version-mark">New in v8.0.0</span>

- Scope: SESSION | GLOBAL
- Persists to cluster: Yes
- Applies to hint [SET_VAR](/optimizer-hints.md#set_varvar_namevar_value): Yes
- Type: Float
- Default value: `-1`
- Range: `[-1, 1]`
- This variable is used to influence the estimated number of rows for an index that matches the SQL statement `ORDER BY` when there are `ORDER BY` and `LIMIT` clauses with filter conditions that are not covered by the index.
- This addresses the same query patterns as the system variable [tidb_opt_ordering_index_selectivity_threshold](#tidb_opt_ordering_index_selectivity_threshold-new-in-v700).
- It differs in implementation by applying a ratio or percentage of the possible range that the qualified rows will be found.

> - Value `-1` (default) and any other value below zero disables this ratio, which enables optimizer to estimate target rows. Any value between `0` and `1` applies a ratio of 0% to 100% (for example, `0.5` = `50%`).

Reviewer: ```suggestion
- Value `-1` (default) and any other value below zero disables this ratio. Any value between `0` and `1` applies a ratio of 0% to 100% (for example, `0.5` = `50%`).
```

| id                                | estRows | task      | access object         | operator info                   |
+-----------------------------------+---------+-----------+-----------------------+---------------------------------+
| Limit_12                          | 1.00    | root      |                       | offset:0, count:1               |
| └─Projection_22                   | 1.00    | root      |                       | test.t.a, test.t.b, test.t.c    |
|   └─IndexLookUp_21                | 1.00    | root      |                       |                                 |
|     ├─IndexFullScan_18(Build)     | 109.20  | cop[tikv] | table:t, index:ia(a)  | keep order:true                 |
|     └─Selection_20(Probe)         | 1.00    | cop[tikv] |                       | le(test.t.b, 9000)              |
|       └─TableRowIDScan_19         | 109.20  | cop[tikv] | table:t               | keep order:false                |
+-----------------------------------+---------+-----------+-----------------------+---------------------------------+
```

- The second example uses `0`, which assumes that 0% of the rows will be scanned before the qualified rows are found. 

> ```sql
> >SET SESSION tidb_opt_ordering_index_selectivity_ratio = 0;
> 
> >EXPLAIN SELECT * FROM t USE INDEX (ia) WHERE b <= 9000 ORDER BY a LIMIT 1;
> +-----------------------------------+---------+-----------+-----------------------+---------------------------------+
> | id                                | estRows | task      | access object         | operator info                   |
> +-----------------------------------+---------+-----------+-----------------------+---------------------------------+
> | Limit_12                          | 1.00    | root      |                       | offset:0, count:1               |
> | └─Projection_22                   | 1.00    | root      |                       | test.t.a, test.t.b, test.t.c    |
> |   └─IndexLookUp_21                | 1.00    | root      |                       |                                 |
> |     ├─IndexFullScan_18(Build)     | 1.00    | cop[tikv] | table:t, index:ia(a)  | keep order:true                 |
> |     └─Selection_20(Probe)         | 1.00    | cop[tikv] |                       | le(test.t.b, 9000)              |
> |       └─TableRowIDScan_19         | 1.00    | cop[tikv] | table:t               | keep order:false                |
> +-----------------------------------+---------+-----------+-----------------------+---------------------------------+
> ```

Reviewer 2: ````suggestion
    ```sql
    > SET SESSION tidb_opt_ordering_index_selectivity_ratio = 0;

    > EXPLAIN SELECT * FROM t USE INDEX (ia) WHERE b <= 9000 ORDER BY a LIMIT 1;
    +-----------------------------------+---------+-----------+-----------------------+---------------------------------+
    | id                                | estRows | task      | access object         | operator info                   |
    +-----------------------------------+---------+-----------+-----------------------+---------------------------------+
    | Limit_12                          | 1.00    | root      |                       | offset:0, count:1               |
    | └─Projection_22                   | 1.00    | root      |                       | test.t.a, test.t.b, test.t.c    |
    |   └─IndexLookUp_21                | 1.00    | root      |                       |                                 |
    |     ├─IndexFullScan_18(Build)     | 1.00    | cop[tikv] | table:t, index:ia(a)  | keep order:true                 |
    |     └─Selection_20(Probe)         | 1.00    | cop[tikv] |                       | le(test.t.b, 9000)              |
    |       └─TableRowIDScan_19         | 1.00    | cop[tikv] | table:t               | keep order:false                |
    +-----------------------------------+---------+-----------+-----------------------+---------------------------------+
    ```
````

- The third example uses `0.1`, meaning that 10% of the possible range is estimated to be scanned. Given the strong filtering of less than 1% of the rows qualified, the worst case is that 99% of the rows need to be scanned before that 1% of the rows are found. 10% of that 99% is approximately 9.9%, which is reflected in the estRows.

> ```sql
> >SET SESSION tidb_opt_ordering_index_selectivity_ratio = 0.1;
> 
> >EXPLAIN SELECT * FROM t USE INDEX (ia) WHERE b <= 9000 ORDER BY a LIMIT 1;
> +-----------------------------------+----------+-----------+-----------------------+---------------------------------+
> | id                                | estRows  | task      | access object         | operator info                   |
> +-----------------------------------+----------+-----------+-----------------------+---------------------------------+
> | Limit_12                          | 1.00     | root      |                       | offset:0, count:1               |
> | └─Projection_22                   | 1.00     | root      |                       | test.t.a, test.t.b, test.t.c    |
> |   └─IndexLookUp_21                | 1.00     | root      |                       |                                 |
> |     ├─IndexFullScan_18(Build)     | 99085.21 | cop[tikv] | table:t, index:ia(a)  | keep order:true                 |
> |     └─Selection_20(Probe)         | 1.00     | cop[tikv] |                       | le(test.t.b, 9000)              |
> |       └─TableRowIDScan_19         | 99085.21 | cop[tikv] | table:t               | keep order:false                |
> +-----------------------------------+----------+-----------+-----------------------+---------------------------------+
> ```

Reviewer 2: ````suggestion
    ```sql
    > SET SESSION tidb_opt_ordering_index_selectivity_ratio = 0.1;

    > EXPLAIN SELECT * FROM t USE INDEX (ia) WHERE b <= 9000 ORDER BY a LIMIT 1;
    +-----------------------------------+----------+-----------+-----------------------+---------------------------------+
    | id                                | estRows  | task      | access object         | operator info                   |
    +-----------------------------------+----------+-----------+-----------------------+---------------------------------+
    | Limit_12                          | 1.00     | root      |                       | offset:0, count:1               |
    | └─Projection_22                   | 1.00     | root      |                       | test.t.a, test.t.b, test.t.c    |
    |   └─IndexLookUp_21                | 1.00     | root      |                       |                                 |
    |     ├─IndexFullScan_18(Build)     | 99085.21 | cop[tikv] | table:t, index:ia(a)  | keep order:true                 |
    |     └─Selection_20(Probe)         | 1.00     | cop[tikv] |                       | le(test.t.b, 9000)              |
    |       └─TableRowIDScan_19         | 99085.21 | cop[tikv] | table:t               | keep order:false                |
    +-----------------------------------+----------+-----------+-----------------------+---------------------------------+
    ```
````

- The fourth example uses `1.0`, which means 100% of the possible range is estimated to be scanned.

> ```sql
> >SET SESSION tidb_opt_ordering_index_selectivity_ratio = 1;
> 
> >EXPLAIN SELECT * FROM t USE INDEX (ia) WHERE b <= 9000 ORDER BY a LIMIT 1;
> +-----------------------------------+-----------+-----------+-----------------------+---------------------------------+
> | id                                | estRows   | task      | access object         | operator info                   |
> +-----------------------------------+-----------+-----------+-----------------------+---------------------------------+
> | Limit_12                          | 1.00      | root      |                       | offset:0, count:1               |
> | └─Projection_22                   | 1.00      | root      |                       | test.t.a, test.t.b, test.t.c    |
> |   └─IndexLookUp_21                | 1.00      | root      |                       |                                 |
> |     ├─IndexFullScan_18(Build)     | 990843.14 | cop[tikv] | table:t, index:ia(a)  | keep order:true                 |
> |     └─Selection_20(Probe)         | 1.00      | cop[tikv] |                       | le(test.t.b, 9000)              |
> |       └─TableRowIDScan_19         | 990843.14 | cop[tikv] | table:t               | keep order:false                |
> +-----------------------------------+-----------+-----------+-----------------------+---------------------------------+
> ```

Reviewer 2: ````suggestion
    ```sql
    > SET SESSION tidb_opt_ordering_index_selectivity_ratio = 1;

    > EXPLAIN SELECT * FROM t USE INDEX (ia) WHERE b <= 9000 ORDER BY a LIMIT 1;
    +-----------------------------------+-----------+-----------+-----------------------+---------------------------------+
    | id                                | estRows   | task      | access object         | operator info                   |
    +-----------------------------------+-----------+-----------+-----------------------+---------------------------------+
    | Limit_12                          | 1.00      | root      |                       | offset:0, count:1               |
    | └─Projection_22                   | 1.00      | root      |                       | test.t.a, test.t.b, test.t.c    |
    |   └─IndexLookUp_21                | 1.00      | root      |                       |                                 |
    |     ├─IndexFullScan_18(Build)     | 990843.14 | cop[tikv] | table:t, index:ia(a)  | keep order:true                 |
    |     └─Selection_20(Probe)         | 1.00      | cop[tikv] |                       | le(test.t.b, 9000)              |
    |       └─TableRowIDScan_19         | 990843.14 | cop[tikv] | table:t               | keep order:false                |
    +-----------------------------------+-----------+-----------+-----------------------+---------------------------------+
    ```
````

- The fifth example also uses `1.0`, but adds a predicate on a that limits the worst case scan range since `WHERE a <= 9000` is matching on the index such that approximately 9000 rows would qualify in total. Given that there is a filtering predicate on `b` that is not in the index, all of the approximately 9000 rows are considered to be scanned before a qualified row for `b <= 9000` is found.

> ```sql
> >EXPLAIN SELECT * FROM t USE INDEX (ia) WHERE a <= 9000 AND b <= 9000 ORDER BY a LIMIT 1;
> +------------------------------------+---------+-----------+-----------------------+------------------------------------+
> | id                                 | estRows | task      | access object         | operator info                      |
> +------------------------------------+---------+-----------+-----------------------+------------------------------------+
> | Limit_12                           | 1.00    | root      |                       | offset:0, count:1                  |
> | └─Projection_22                    | 1.00    | root      |                       | test.t.a, test.t.b, test.t.c       |
> |   └─IndexLookUp_21                 | 1.00    | root      |                       |                                    |
> |     ├─IndexRangeScan_18(Build)     | 9074.99 | cop[tikv] | table:t, index:ia(a)  | range:[-inf,9000], keep order:true |
> |     └─Selection_20(Probe)          | 1.00    | cop[tikv] |                       | le(test.t.b, 9000)                 |
> |       └─TableRowIDScan_19          | 9074.99 | cop[tikv] | table:t               | keep order:false                   |
> +------------------------------------+---------+-----------+-----------------------+------------------------------------+
> ```

Reviewer 2: ````suggestion
    ```sql
    > SET SESSION tidb_opt_ordering_index_selectivity_ratio = 1;

    > EXPLAIN SELECT * FROM t USE INDEX (ia) WHERE a <= 9000 AND b <= 9000 ORDER BY a LIMIT 1;
    +------------------------------------+---------+-----------+-----------------------+------------------------------------+
    | id                                 | estRows | task      | access object         | operator info                      |
    +------------------------------------+---------+-----------+-----------------------+------------------------------------+
    | Limit_12                           | 1.00    | root      |                       | offset:0, count:1                  |
    | └─Projection_22                    | 1.00    | root      |                       | test.t.a, test.t.b, test.t.c       |
    |   └─IndexLookUp_21                 | 1.00    | root      |                       |                                    |
    |     ├─IndexRangeScan_18(Build)     | 9074.99 | cop[tikv] | table:t, index:ia(a)  | range:[-inf,9000], keep order:true |
    |     └─Selection_20(Probe)          | 1.00    | cop[tikv] |                       | le(test.t.b, 9000)                 |
    |       └─TableRowIDScan_19          | 9074.99 | cop[tikv] | table:t               | keep order:false                   |
    +------------------------------------+---------+-----------+-----------------------+------------------------------------+
    ```
````

- The real-time statistics are the total number of rows and the number of modified rows that are automatically updated based on DML statements. When this variable is set to `moderate` (default), TiDB generates the execution plan based on real-time statistics. When this variable is set to `determinate`, TiDB does not use real-time statistics for generating the execution plan, which will make execution plans more stable.
- For long-term stable OLTP workload, or if the user is affirmative on the existing execution plans, it is recommended to use the `determinate` mode to reduce the possibility of unexpected execution plan changes. Additionally, you can use the [`LOCK STATS`](/sql-statements/sql-statement-lock-stats.md) to prevent the statistics from being modified and further stabilize the execution plan.

### tidb_opt_ordering_index_selectivity_ratio <span class="version-mark">New in v8.0.0</span>

- Scope: SESSION | GLOBAL
- Persists to cluster: Yes
- Applies to hint [SET_VAR](/optimizer-hints.md#set_varvar_namevar_value): Yes
- Type: Float
- Default value: `-1`
- Range: `[-1, 1]`

> - This variable controls the estimated number of rows for an index that matches the SQL statement `ORDER BY` when there are `ORDER BY` and `LIMIT` clauses in a SQL statement, but does not cover some filter conditions.

Reviewer 2: ```suggestion
- This variable controls the estimated number of rows for an index that matches the SQL statement `ORDER BY` when there are `ORDER BY` and `LIMIT` clauses in a SQL statement, but some filter conditions are not covered by the index.
```

> - This variable is used to influence the estimated number of rows for an index that matches the SQL statement `ORDER BY` when there are `ORDER BY` and `LIMIT` clauses with filter conditions that are not covered by the index.

Reviewer: ```suggestion
- This variable is used to influence the estimated number of rows for an index that matches the SQL statement `ORDER BY` when there are `ORDER BY` and `LIMIT` clauses but the filter conditions are not covered by the index.
```

> - This variable is used to influence the estimated number of rows for an index that matches the SQL statement `ORDER BY` when there are `ORDER BY` and `LIMIT` clauses with filter conditions that are not covered by the index.

Reviewer 2: ```suggestion
- This variable controls the estimated number of rows for an index that matches the `ORDER BY` and `LIMIT` clauses in a SQL statement, but does not cover some filter conditions.
```

> - This addresses the same query patterns as the system variable [tidb_opt_ordering_index_selectivity_threshold](#tidb_opt_ordering_index_selectivity_threshold-new-in-v700).
> - It differs in implementation by applying a ratio or percentage of the possible range that the qualified rows will be found.

Reviewer 2: ```suggestion
- This variable is applicable in scenarios similar to those of the system variable [`tidb_opt_ordering_index_selectivity_threshold`](#tidb_opt_ordering_index_selectivity_threshold-new-in-v700).
- Unlike the implementation of `tidb_opt_ordering_index_selectivity_threshold`, this variable applies a ratio or percentage of the possible number of rows that meet the conditions within the range.
```

- A value of `-1` (default) or less thabn `0` disables this ratio. Any value between `0` and `1` applies a ratio of 0% to 100% (for example, `0.5` corresponds to `50%`).

> - In the following examples, the table `t` has a total of 1,000,000 rows. The same query is used, but different values for `tidb_opt_ordering_index_selectivity_ratio` are used. The query in the example contains a `WHERE` clause predicate that qualifies a small percentage of the rows (9,000 out of 1,000,000). There is an index that supports the `ORDER BY a` (index `ia`), but the filter on `b` is not included in this index. Depending on the actual data distribution, the rows matching the `WHERE` clause and `LIMIT 1` might be found as the first row accessed when scanning the non-filtering index, or at worst, after nearly all of the rows have been processed.

Reviewer 2: ```suggestion
- In the following examples, the table `t` has a total of 1,000,000 rows. The same query is used, but different values for `tidb_opt_ordering_index_selectivity_ratio` are used. The query in the example contains a `WHERE` clause predicate that qualifies a small percentage of rows (9,000 out of 1,000,000). There is an index that supports the `ORDER BY a` (index `ia`), but the filter on `b` is not included in this index. Depending on the actual data distribution, the rows matching the `WHERE` clause and `LIMIT 1` might be found as the first row accessed when scanning the non-filtering index, or at worst, after nearly all the rows have been processed.
```

> - Value `-1` (default) and any other value below zero disables this ratio. Any value between `0` and `1` applies a ratio of 0% to 100% (for example, `0.5` = `50%`).
> - In the following examples, the table `t` has a total of 1,000,000 rows. The same query is used, but different values for `tidb_opt_ordering_index_selectivity_ratio` are used. The query in the example has a `WHERE` clause predicate that qualifies a small percentage of the rows (9000 out of 1,000,000). There is an index that supports the `ORDER BY a` (index `ia`), but the filtering on `b` does not appear in this index. Based upon the data distribution, the row matching the `WHERE` clause and `LIMIT 1` can be found as the 1st row accessed when scanning the non-filtering index, or at worst, after nearly all of the rows have been processed.
> - An index hint is used in each example to demonstrate the impact on estRows. The impact on the final plan choice is dependent on the availability and cost of other plans.

Reviewer 2: ```suggestion
- A value of `-1` (default) or less than `0` disables this ratio. Any value between `0` and `1` applies a ratio of 0% to 100% (for example, `0.5` corresponds to  `50%`).
- In the following examples, the table `t` has a total of 1,000,000 rows. The same query is used, but different values for `tidb_opt_ordering_index_selectivity_ratio` are used. The query in the example contains a `WHERE` clause predicate that qualifies a small percentage of rows (9,000 out of 1,000,000). There is an index that supports the `ORDER BY a` (index `ia`), but the filter on `b` is not included in this index. Depending on the actual data distribution, the rows matching the `WHERE` clause and `LIMIT 1` might be found as the first row accessed when scanning the non-filtering index, or at worst, after nearly all the rows have been processed.
- Each example uses an index hint to demonstrate the impact on estRows. The final plan selection depends on the availability and cost of other plans.
```

> - The first example uses the default value `-1`, which uses the existing estimation formula. The default behavior is that a small percentage of the rows are estimated to be scanned before a row is found that qualifies from filtering outside of that index.

Reviewer 2: ```suggestion
- The first example uses the default value `-1`, which uses the existing estimation formula. By default, a small percentage of rows are scanned for estimation before the qualified rows are found.
```

    | id                                | estRows | task      | access object         | operator info                   |
    +-----------------------------------+---------+-----------+-----------------------+---------------------------------+
    | Limit_12                          | 1.00    | root      |                       | offset:0, count:1               |
    | └─Projection_22                   | 1.00    | root      |                       | test.t.a, test.t.b, test.t.c    |
    |   └─IndexLookUp_21                | 1.00    | root      |                       |                                 |
    |     ├─IndexFullScan_18(Build)     | 109.20  | cop[tikv] | table:t, index:ia(a)  | keep order:true                 |
    |     └─Selection_20(Probe)         | 1.00    | cop[tikv] |                       | le(test.t.b, 9000)              |
    |       └─TableRowIDScan_19         | 109.20  | cop[tikv] | table:t               | keep order:false                |
    +-----------------------------------+---------+-----------+-----------------------+---------------------------------+
    ```

> - The second example uses `0`, which assumes that 0% of the rows will be scanned before the qualified rows are found. 

Reviewer 2: ```suggestion
- The second example uses `0`, which assumes that 0% of rows will be scanned before the qualified rows are found. 
```

    | id                                | estRows | task      | access object         | operator info                   |
    +-----------------------------------+---------+-----------+-----------------------+---------------------------------+
    | Limit_12                          | 1.00    | root      |                       | offset:0, count:1               |
    | └─Projection_22                   | 1.00    | root      |                       | test.t.a, test.t.b, test.t.c    |
    |   └─IndexLookUp_21                | 1.00    | root      |                       |                                 |
    |     ├─IndexFullScan_18(Build)     | 1.00    | cop[tikv] | table:t, index:ia(a)  | keep order:true                 |
    |     └─Selection_20(Probe)         | 1.00    | cop[tikv] |                       | le(test.t.b, 9000)              |
    |       └─TableRowIDScan_19         | 1.00    | cop[tikv] | table:t               | keep order:false                |
    +-----------------------------------+---------+-----------+-----------------------+---------------------------------+
    ```

> - The third example uses `0.1`, meaning that 10% of the possible range is estimated to be scanned. Given the strong filtering of less than 1% of the rows qualified, the worst case is that 99% of the rows need to be scanned before that 1% of the rows are found. 10% of that 99% is approximately 9.9%, which is reflected in the estRows.

Reviewer 2: ```suggestion
- The third example uses `0.1`, which assumes that 10% of rows will be scanned before the qualified rows are found. This condition is highly selective, with only 1% of rows meeting the condition. Therefore, in the worst-case scenario, it might be necessary to scan 99% of rows before finding the 1% that qualify. 10% of that 99% is approximately 9.9%, which is reflected in the estRows.
```

    | id                                | estRows  | task      | access object         | operator info                   |
    +-----------------------------------+----------+-----------+-----------------------+---------------------------------+
    | Limit_12                          | 1.00     | root      |                       | offset:0, count:1               |
    | └─Projection_22                   | 1.00     | root      |                       | test.t.a, test.t.b, test.t.c    |
    |   └─IndexLookUp_21                | 1.00     | root      |                       |                                 |
    |     ├─IndexFullScan_18(Build)     | 99085.21 | cop[tikv] | table:t, index:ia(a)  | keep order:true                 |
    |     └─Selection_20(Probe)         | 1.00     | cop[tikv] |                       | le(test.t.b, 9000)              |
    |       └─TableRowIDScan_19         | 99085.21 | cop[tikv] | table:t               | keep order:false                |
    +-----------------------------------+----------+-----------+-----------------------+---------------------------------+
    ```

> - The fourth example uses `1.0`, which means 100% of the possible range is estimated to be scanned.

Reviewer 2: ```suggestion
- The fourth example uses `1.0`, which assumes that 100% of rows will be scanned before the qualified rows are found.
```

    | id                                | estRows   | task      | access object         | operator info                   |
    +-----------------------------------+-----------+-----------+-----------------------+---------------------------------+
    | Limit_12                          | 1.00      | root      |                       | offset:0, count:1               |
    | └─Projection_22                   | 1.00      | root      |                       | test.t.a, test.t.b, test.t.c    |
    |   └─IndexLookUp_21                | 1.00      | root      |                       |                                 |
    |     ├─IndexFullScan_18(Build)     | 990843.14 | cop[tikv] | table:t, index:ia(a)  | keep order:true                 |
    |     └─Selection_20(Probe)         | 1.00      | cop[tikv] |                       | le(test.t.b, 9000)              |
    |       └─TableRowIDScan_19         | 990843.14 | cop[tikv] | table:t               | keep order:false                |
    +-----------------------------------+-----------+-----------+-----------------------+---------------------------------+
    ```

> - The fifth example also uses `1.0`, but adds a predicate on a that limits the worst case scan range since `WHERE a <= 9000` is matching on the index such that approximately 9000 rows would qualify in total. Given that there is a filtering predicate on `b` that is not in the index, all of the approximately 9000 rows are considered to be scanned before a qualified row for `b <= 9000` is found.

Reviewer 2: ```suggestion
- The fifth example also uses `1.0`, but adds a predicate on `a`, limiting the scan range in the worst-case scenario. This is because `WHERE a <= 9000` matches the index, with approximately 9,000 rows would qualify. Given that the filter predicate on `b` is not in the index, all the approximately 9,000 rows are considered to be scanned before finding a row that matches `b <= 9000`.
```

- Type: Float
- Default value: `-1`
- Range: `[-1, 1]`
- This variable controls the estimated number of rows for an index that matches the SQL statement `ORDER BY` when there are `ORDER BY` and `LIMIT` clauses in a SQL statement, but does not cover some filter conditions.
- This addresses the same query patterns as the system variable [tidb_opt_ordering_index_selectivity_threshold](#tidb_opt_ordering_index_selectivity_threshold-new-in-v700).
- It differs in implementation by applying a ratio or percentage of the possible range that the qualified rows will be found.

> - A value of `-1` (default) or less thabn `0` disables this ratio. Any value between `0` and `1` applies a ratio of 0% to 100% (for example, `0.5` corresponds to `50%`).

Reviewer 2: ```suggestion
- A value of `-1` (default) or less than `0` disables this ratio. Any value between `0` and `1` applies a ratio of 0% to 100% (for example, `0.5` corresponds to `50%`).
```

+----------------------------------+---------+-----------+----------------------+-------------------------------------+
```

### tidb_opt_ordering_index_selectivity_ratio <span class="version-mark">New in v8.0.0</span>

- Scope: SESSION | GLOBAL
- Persists to cluster: Yes
- Applies to hint [SET_VAR](/optimizer-hints.md#set_varvar_namevar_value): Yes
- Type: Float
- Default value: `-1`
- Range: `[-1, 1]`
- This variable is used to influence the estimated number of rows for an index that matches a SQL statements `ORDER BY` when there are `ORDER BY` and `LIMIT` clauses with filter conditions that aren't covered by the index.

> - This addresses the same query patterns as variable [tidb_opt_ordering_index_selectivity_threshold](#tidb_opt_ordering_index_selectivity_threshold-new-in-v700).

Reviewer 3: If we set both threshold and ratio, which one take effects?

Reviewer 4: From my understanding, if the query satisfies the threshold, the ratio won't and also doesn't need to take effect. In that case, the effect will be similar to setting the ratio to `1.0`.

- It differs in implementation by applying a ratio or percentage of the possible range that the qualified rows will be found.

> - Value -1 (default) disables this ratio. Any value between 0 and 1 applies a ratio of 0% to 100% (meaning 0.5 = 50%).

Reviewer 3: ```suggestion
- Value -1 (default) disables this ratio, that enables optimizer to estimate target rows. Any value between 0 and 1 applies a ratio of 0% to 100% (meaning 0.5 = 50%).
```

> - Value -1 (default) disables this ratio. Any value between 0 and 1 applies a ratio of 0% to 100% (meaning 0.5 = 50%).

Reviewer 3: ```suggestion
- Value -1 (default) disables this ratio, that enables optimizer to estimate target rows. Any value between 0 and 1 applies a ratio of 0% to 100% (meaning 0.5 = 50%).
```

Reviewer 3: What if we set it to a value (-1,0) ?

Author: Any value below zero has the same effect as -1. I have added this to the doc.

- In the following examples, table `t` has a total of 1,000,000 rows. The same query is used, but different values for `tidb_opt_ordering_index_selectivity_ratio` are used. The query in the example has a WHERE clause predicate that qualifies a small percentage of the rows (9000 out of 1,000,000). There is an index that supports the `ORDER BY a` (index ia), but the filtering on b does not appear in this index. Based upon the data distribution, the row matching the WHERE clause and LIMIT 1 could be found as the 1st row accessed when scanning the non-filtering index, or at worst, after nearly all of the rows have been processed.
- The first example uses the default -1, which uses the existing estimation formula. The default behavior is that a small percentage of the rows are estimated to be scanned before a row is found that qualifies from filtering outside of that index.

```sql
> SET SESSION tidb_opt_ordering_index_selectivity_ratio = -1;

> >EXPLAIN SELECT * FROM t USE INDEX (ia) WHERE b <= 9000 ORDER BY a 1;

Reviewer 4: The `LIMIT` is missing. And the same for the SQLs below.

Author: Done. Thank you. I appreciate you catching this.

+-----------------------------------+---------+-----------+-----------------------+---------------------------------+
| Limit_12                          | 1.00    | root      |                       | offset:0, count:1               |
| └─Projection_22                   | 1.00    | root      |                       | test.t.a, test.t.b, test.t.c    |
|   └─IndexLookUp_21                | 1.00    | root      |                       |                                 |
|     ├─IndexFullScan_18(Build)     | 109.20  | cop[tikv] | table:t, index:ia(a)  | keep order:true                 |
|     └─Selection_20(Probe)         | 1.00    | cop[tikv] |                       | le(test.t.b, 9000)              |
|       └─TableRowIDScan_19         | 109.20  | cop[tikv] | table:t               | keep order:false                |
+-----------------------------------+---------+-----------+-----------------------+---------------------------------+
```

- The 2nd example uses 0, which assumes that 0% of the rows will be scanned before the qualified rows are found. 

```sql
>SET SESSION tidb_opt_ordering_index_selectivity_ratio = 0;

> >EXPLAIN SELECT * FROM t USE INDEX (ia) WHERE b <= 9000 ORDER BY a 1;

Reviewer 3: I assume the reason we need this rectification is to make better decision of table/index access. 
In the examples, forcing index access doesn't reflect the decision change caused by the variable.  Is it a better idea if we remove the hints?  The different values switch table/index access.

Author: The hints are there to demonstrate the effect on the estimated rows. Whether this new estimate results in that index losing to another plan is entirely dependent on the cost of other plans.

+----------------------------------+---------+-----------+----------------------+-------------------------------------+
```

### tidb_opt_ordering_index_selectivity_ratio <span class="version-mark">New in v8.0.0</span>

- Scope: SESSION | GLOBAL
- Persists to cluster: Yes
- Applies to hint [SET_VAR](/optimizer-hints.md#set_varvar_namevar_value): Yes
- Type: Float
- Default value: `-1`
- Range: `[-1, 1]`

> - This variable is used to influence the estimated number of rows for an index that matches a SQL statements `ORDER BY` when there are `ORDER BY` and `LIMIT` clauses with filter conditions that aren't covered by the index.

Reviewer: ```suggestion
- This variable is used to influence the estimated number of rows for an index that matches the SQL statement `ORDER BY` when there are `ORDER BY` and `LIMIT` clauses with filter conditions that are not covered by the index.
```

- This addresses the same query patterns as variable [tidb_opt_ordering_index_selectivity_threshold](#tidb_opt_ordering_index_selectivity_threshold-new-in-v700).
- It differs in implementation by applying a ratio or percentage of the possible range that the qualified rows will be found.

> - Value -1 (default) and any other value below zero disables this ratio, that enables optimizer to estimate target rows. Any value between 0 and 1 applies a ratio of 0% to 100% (meaning 0.5 = 50%).

Reviewer: ```suggestion
- Value `-1` (default) and any other value below zero disables this ratio, which enables optimizer to estimate target rows. Any value between `0` and `1` applies a ratio of 0% to 100% (for example, `0.5` means `50%`).
```

- In the following examples, table `t` has a total of 1,000,000 rows. The same query is used, but different values for `tidb_opt_ordering_index_selectivity_ratio` are used. The query in the example has a WHERE clause predicate that qualifies a small percentage of the rows (9000 out of 1,000,000). There is an index that supports the `ORDER BY a` (index ia), but the filtering on b does not appear in this index. Based upon the data distribution, the row matching the WHERE clause and LIMIT 1 could be found as the 1st row accessed when scanning the non-filtering index, or at worst, after nearly all of the rows have been processed.
- An index hint is used in each example to demonstrate the impact to estRows. The impact to the final plan choice is dependent on the availability and cost of other plans.

> - The first example uses the default -1, which uses the existing estimation formula. The default behavior is that a small percentage of the rows are estimated to be scanned before a row is found that qualifies from filtering outside of that index.

Reviewer: ```suggestion
- The first example uses the default value `-1`, which uses the existing estimation formula. The default behavior is that a small percentage of the rows are estimated to be scanned before a row is found that qualifies from filtering outside of that index.
```

+-----------------------------------+----------+-----------+-----------------------+---------------------------------+
| id                                | estRows  | task      | access object         | operator info                   |
+-----------------------------------+----------+-----------+-----------------------+---------------------------------+
| Limit_12                          | 1.00     | root      |                       | offset:0, count:1               |
| └─Projection_22                   | 1.00     | root      |                       | test.t.a, test.t.b, test.t.c    |
|   └─IndexLookUp_21                | 1.00     | root      |                       |                                 |
|     ├─IndexFullScan_18(Build)     | 99085.21 | cop[tikv] | table:t, index:ia(a)  | keep order:true                 |
|     └─Selection_20(Probe)         | 1.00     | cop[tikv] |                       | le(test.t.b, 9000)              |
|       └─TableRowIDScan_19         | 99085.21 | cop[tikv] | table:t               | keep order:false                |
+-----------------------------------+----------+-----------+-----------------------+---------------------------------+
```

> - The 4th example uses 1.0, which means 100% of the possible range is estimated to be scanned.

Reviewer: ```suggestion
- The fourth example uses `1`, which means 100% of the possible range is estimated to be scanned.
```

| id                                | estRows   | task      | access object         | operator info                   |
+-----------------------------------+-----------+-----------+-----------------------+---------------------------------+
| Limit_12                          | 1.00      | root      |                       | offset:0, count:1               |
| └─Projection_22                   | 1.00      | root      |                       | test.t.a, test.t.b, test.t.c    |
|   └─IndexLookUp_21                | 1.00      | root      |                       |                                 |
|     ├─IndexFullScan_18(Build)     | 990843.14 | cop[tikv] | table:t, index:ia(a)  | keep order:true                 |
|     └─Selection_20(Probe)         | 1.00      | cop[tikv] |                       | le(test.t.b, 9000)              |
|       └─TableRowIDScan_19         | 990843.14 | cop[tikv] | table:t               | keep order:false                |
+-----------------------------------+-----------+-----------+-----------------------+---------------------------------+
```

> - The 5th example also uses 1.0, but adds a predicate on a that limits the worst case scan range since `WHERE a <= 9000` is matching on the index such that approximately 9000 rows would qualify in total. Given that there is a filtering predicate on b that is not in the index, all of the approximately 9000 rows are considered to be scanned before a qualified row for `b <= 9000` is found.

Reviewer: ```suggestion
- The fifth example also uses `1`, but adds a predicate on `a` that limits the worst case scan range since `WHERE a <= 9000` is matching on the index so that approximately 9000 rows would qualify in total. Given that there is a filtering predicate on `b` that is not in the index, all of the approximately 9000 rows are considered to be scanned before a qualified row for `b <= 9000` is found.
```
